types/leaflet/index.d.ts

```diff
@@ -424,11 +424,11 @@ declare namespace L {
         getTooltip(): Tooltip | undefined;
 
         // Extension methods
-        onAdd?: (map: Map) => this;
-        onRemove?: (map: Map) => this;
-        getEvents?: () => {[name: string]: (event: Event) => void};
-        getAttribution?: () => string | null;
-        beforeAdd?: (map: Map) => this;
+        onAdd?(map: Map): this;
```

Comment:
I have marked them as optional on `Layer` because they aren't present on that class (which means that you correctly get a warning if you call super.onAdd on a class extending `Layer`). 

A solution would be to declare in each class extending `Layer`, which of the extension methods are defined (this would include declaring `onAdd` on the `Marker` class). But this just seems like to much redundant information. 

Marking the class and the extension methods as abstract just forces the classes extending layer to redundantly declare them. 

I'll remove the optional annotation from the non-optional extension methods in `L.Marker` .